craft_archives/repo/package_repository.py

```diff
@@ -45,15 +54,64 @@ class PriorityString(enum.IntEnum):
 PriorityValue = Union[int, Literal["always", "prefer", "defer"]]
 
 
-class PackageRepository(abc.ABC):
+class PackageRepository(pydantic.BaseModel, abc.ABC):
     """The base class for package repositories."""
 
-    @abc.abstractmethod
-    def marshal(self) -> Dict[str, Any]:
+    class Config:  # pylint: disable=too-few-public-methods
+        """Pydantic model configuration."""
+
+        # pyright: reportUnknownMemberType=false
+        # pyright: reportUnknownVariableType=false
+        # pyright: reportUnknownLambdaType=false
+
+        validate_assignment = True
+        allow_mutation = False
+        allow_population_by_field_name = True
+        alias_generator = lambda s: s.replace("_", "-")  # noqa: E731
+        extra = "forbid"
+
+    type: Literal["apt"]
```

Comment:
Not something we have to address, but I glanced at the changes you made for snapcraft, and defaulting this value to "apt" here (or as overrides in the two concrete classes) would ensure complete backwards compatibility if we want that.

Reply:
I thought about doing that, but wouldn't it mean that users (the ones writing the `snapcraft.yaml` or `rockcraft.yaml`) would be able to declare package repositories _without_ specifying the type?